beacon-chain/core/state/transition.go

```diff
@@ -14,40 +15,62 @@ import (
 	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
 )
 
+var log = logrus.WithField("prefix", "core/state")
+
 // ExecuteStateTransition defines the procedure for a state transition function.
 // Spec:
-//  We now define the state transition function. At a high level the state transition is made up of two parts:
-//  - The per-slot transitions, which happens every slot, and only affects a parts of the state.
-//  - The per-epoch transitions, which happens at every epoch boundary (i.e. state.slot % SLOTS_PER_EPOCH == 0), and affects the entire state.
-//  The per-slot transitions generally focus on verifying aggregate signatures and saving temporary records relating to the per-slot
-//  activity in the BeaconState. The per-epoch transitions focus on the validator registry, including adjusting balances and activating
-//  and exiting validators, as well as processing crosslinks and managing block justification/finalization.
+//  We now define the state transition function. At a high level the state transition is made up of three parts:
+//  - The per-slot transitions, which happens at the start of every slot.
+//  - The per-block transitions, which happens at every block.
+//  - The per-epoch transitions, which happens at the end of the last slot of every epoch (i.e. (state.slot + 1) % SLOTS_PER_EPOCH == 0).
+//  The per-slot transitions focus on the slot counter and block roots records updates.
+//  The per-block transitions focus on verifying aggregate signatures and saving temporary records relating to the per-block activity in the state.
+//  The per-epoch transitions focus on the validator registry, including adjusting balances and activating and exiting validators,
+//  as well as processing crosslinks and managing block justification/finalization.
 func ExecuteStateTransition(
-	beaconState *pb.BeaconState,
+	state *pb.BeaconState,
 	block *pb.BeaconBlock,
-	prevBlockRoot [32]byte,
+	headRoot [32]byte,
 	verifySignatures bool,
 ) (*pb.BeaconState, error) {
 	var err error
 
-	currentSlot := beaconState.Slot
-	beaconState.Slot = currentSlot + 1
-	beaconState = b.ProcessBlockRoots(beaconState, prevBlockRoot)
+	// Execute per slot transition.
+	state = ProcessSlot(state, headRoot)
+
+	// Execute per block transition.
 	if block != nil {
-		beaconState, err = ProcessBlock(beaconState, block, verifySignatures)
+		state, err = ProcessBlock(state, block, verifySignatures)
 		if err != nil {
-			return nil, fmt.Errorf("unable to process block: %v", err)
+			return nil, fmt.Errorf("could not process block: %v", err)
 		}
+	}
 
-		if e.CanProcessEpoch(beaconState) {
```

Comment:
This is a bug, `ProcessEpoch` should not be based on the incoming block, it should be based on the slot. If there's no incoming back, epoch will never get a update : )

Reply:
I moved it outside of `if block != nil`